### PR TITLE
updated exchange service, added top 4 exchange currency, sellbuy methods

### PR DESCRIPTION
**Pull request by randrii**

### CODE REVIEW

File: src/main/java/com/rybka/constant/PopularExchangeCurrency.java [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package com.rybka.constant;
+
+public enum PopularExchangeCurrency {
+    UAH, USD, EUR, PLN
```
Comment (jarc0der) [inline src/main/java/com/rybka/constant/PopularExchangeCurrency.java]: Could we provide this list in the app.properties?

File: src/main/java/com/rybka/service/exchange/ExchangeService.java [hunk 2]
```diff
@@ -18,19 +25,47 @@
...
+        if (exchangeType.equals("SELL")) {
+            sellBaseCurrency(count, exchangeResponse, topCurrencyList);
+        } else if (exchangeType.equals("BUY")) {
+            buyBaseCurrency(base, count, topCurrencyList);
+        } else throw new IncorrectUserDataException("Unknown exchange type.");
+
+        return new ExchangeResultData(base, count, topCurrencyList);
+    }
+
+    private void buyBaseCurrency(String base, double count, List<TopCurrencyData> topCurrencyList) {
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/ExchangeService.java]: return a list 
use java8 streams

File: src/main/java/com/rybka/service/exchange/ExchangeService.java [hunk 3]
```diff
@@ -18,19 +25,47 @@
...
     public Double calculateTotal(Double rate, Double amount) {
         return rate * amount;
     }
 
-    private Map.Entry<String, Double> retrieveTargetCurrency(ExchangeResponse exchangeResponse, String userTargetCurrency) {
+    public ExchangeResultData retrieveTopCurrency(String base, String exchangeType, double count) {
+        var exchangeResponse = connectorService.retrieveRates(base);
+        var topCurrencyList = new ArrayList<TopCurrencyData>();
+
+        if (exchangeType.equals("SELL")) {
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/ExchangeService.java]: Try to use map instead of if/else

Comment (jarc0der) [reply src/main/java/com/rybka/service/exchange/ExchangeService.java]: Map<String, BiFunction<Int, ExchangeResponse, List<TopCurrency>>

@PostConstruct <- read about this 
map.add('sell', sellHandler);
map.add('buy' buyHandler);

var topCurrencyList = mapExchangeTypeHandlers.get(exchangeType).accept(count, exchangeResponse);

File: src/main/java/com/rybka/service/exchange/ExchangeService.java [hunk 4]
```diff
@@ -18,19 +25,47 @@
...
+
+            var currencyRate = retrieveTargetCurrency(exchangeResponse1, base).map(Map.Entry::getValue).orElse(0.0);
+            topCurrencyList.add(new TopCurrencyData(currency.name(), currencyRate, calculateTotal(currencyRate, count)));
+        }
+    }
+
+    private void sellBaseCurrency(double count, ExchangeResponse exchangeResponse, List<TopCurrencyData> topCurrencyList) {
+        for (PopularExchangeCurrency currency : PopularExchangeCurrency.values()) {
+            var currencyRate = retrieveTargetCurrency(exchangeResponse, currency.name()).map(Map.Entry::getValue).orElse(0.0);
+            topCurrencyList.add(new TopCurrencyData(currency.name(), currencyRate, calculateTotal(currencyRate, count)));
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/ExchangeService.java]: Check Lombok builders. 

TopCurrencyData.builder().currencyName(currency.name()).currenctyRate(currencyRate);

File: src/main/java/com/rybka/service/exchange/ExchangeService.java [hunk 5]
```diff
@@ -18,19 +25,47 @@
...
+        }
+    }
+
+    private Optional<Map.Entry<String, Double>> retrieveTargetCurrency(ExchangeResponse exchangeResponse, String userTargetCurrency) {
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/ExchangeService.java:95]: Check if we need <key, value> mb we just need value.

File: src/main/java/com/rybka/service/exchange/ExchangeService.java (hunk 3 above)
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/ExchangeService.java]: SELL, BUY - constants.

File: src/main/java/com/rybka/model/ExchangeResultData.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.ToString;
+
+import java.util.List;
+
+@Data
+@AllArgsConstructor
+@ToString
```
Comment (jarc0der) [inline src/main/java/com/rybka/model/ExchangeResultData.java]: Do you need @ToString anotation?

File: src/main/java/com/rybka/service/exchange/ExchangeService.java (hunk 2 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/ExchangeService.java]: fixed

File: src/main/java/com/rybka/constant/PopularExchangeCurrency.java (hunk 1 above)
Comment (randrii) [reply src/main/java/com/rybka/constant/PopularExchangeCurrency.java]: fixed

File: src/main/java/com/rybka/service/exchange/ExchangeService.java (hunk 5 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/ExchangeService.java:95]: Only one requires key and value, other -  only value.

File: src/main/java/com/rybka/model/ExchangeResultData.java (hunk 6 above)
Comment (randrii) [reply src/main/java/com/rybka/model/ExchangeResultData.java]: fixed

File: src/main/java/com/rybka/service/exchange/ExchangeService.java (hunk 3 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/ExchangeService.java]: fixed

File: src/main/java/com/rybka/model/ExchangeResultData.java (hunk 6 above)
Comment (randrii) [reply src/main/java/com/rybka/model/ExchangeResultData.java:11]: fixed

File: src/main/java/com/rybka/service/exchange/ExchangeService.java (hunk 3 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/ExchangeService.java]: fixed

File: src/main/java/com/rybka/service/exchange/ExchangeService.java (hunk 4 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/ExchangeService.java]: fixed

File: src/main/java/com/rybka/service/exchange/ExchangeService.java [hunk 7]
```diff
@@ -1,36 +1,100 @@
...
+        exchangeTypeMap.put(ExchangeType.BUY.name(), this::buyBaseCurrency);
+        exchangeTypeMap.put(ExchangeType.SELL.name(), this::sellBaseCurrency);
+    }
+
+    private List<TopCurrencyData> buyBaseCurrency(String base, double count) {
+        var popularCurrencyList = Objects.requireNonNull(environment.getProperty(PropertyInfo.PROPERTY_EXCHANGE_CURRENCY)).split(",");
+        var topCurrencyList = new ArrayList<TopCurrencyData>();
+
+        Arrays.stream(popularCurrencyList)
+                .forEach(
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/ExchangeService.java]: try to use Map, Collect functions.

Comment (jarc0der) [reply src/main/java/com/rybka/service/exchange/ExchangeService.java]: Method is to big, add additional method to make current smaller. 

Arrays.stream(...).map(s -> retrieveTopCurrency(s)).collect(Collectors.toList());

File: src/main/java/com/rybka/service/exchange/ExchangeService.java [hunk 8]
```diff
@@ -1,36 +1,100 @@
...
+    }
+
+    private List<TopCurrencyData> sellBaseCurrency(String base, double count) {
+        var exchangeResponse = connectorService.retrieveRates(base);
+
+        var popularCurrencyList = Objects.requireNonNull(environment.getProperty(PropertyInfo.PROPERTY_EXCHANGE_CURRENCY)).split(",");
+        var topCurrencyList = new ArrayList<TopCurrencyData>();
+
+        Arrays.stream(popularCurrencyList).forEach(
+                s -> {
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/ExchangeService.java]: The same here. Simplify methods.

File: src/main/java/com/rybka/service/exchange/ExchangeService.java (hunk 7 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/ExchangeService.java]: fixed

File: src/main/java/com/rybka/service/exchange/ExchangeService.java (hunk 8 above)
Comment (randrii) [reply src/main/java/com/rybka/service/exchange/ExchangeService.java]: fixed

File: src/main/java/com/rybka/service/exchange/ExchangeService.java [hunk 9]
```diff
@@ -53,42 +54,42 @@ private void initExchangeTypeMap() {
...
-                });
-
-        return topCurrencyList;
+
+        return Arrays.stream(popularCurrencyList)
+                .map(s -> obtainSellBaseForTopCurrency(s, base, count))
+                .collect(Collectors.toList());
+    }
+
+    private TopCurrencyData obtainBuyBaseForTopCurrency(String s, String base, double count) {
```
Comment (jarc0der) [inline src/main/java/com/rybka/service/exchange/ExchangeService.java]: String s -> parameter name is too bad :(

Comment (randrii) [reply src/main/java/com/rybka/service/exchange/ExchangeService.java]: fixed

**Pull request merged**